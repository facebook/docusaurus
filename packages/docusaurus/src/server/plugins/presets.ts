/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {createRequire} from 'module';
import {loadFreshModule} from '@docusaurus/utils';
import {resolveModuleName} from './moduleShorthand';
import type {
  LoadContext,
  PresetConfigDefined,
  PresetModule,
  Preset,
  DocusaurusConfig,
} from '@docusaurus/types';

type ImportedPresetModule = PresetModule & {default?: PresetModule};

/**
 * Calls preset functions, aggregates each of their return values, and returns
 * the plugin and theme configs.
 */
export async function loadPresets(
  context: LoadContext,
): Promise<Pick<DocusaurusConfig, 'plugins' | 'themes'>> {
  // We need to resolve plugins from the perspective of the site config, as if
  // we are using `require.resolve` on those module names.
  const presetRequire = createRequire(context.siteConfigPath);

  const presets = context.siteConfig.presets.filter(
    (p): p is PresetConfigDefined => !!p,
  );

  async function loadPreset(presetItem: PresetConfigDefined): Promise<Preset> {
    let presetModuleImport: string;
    let presetOptions = {};
    if (typeof presetItem === 'string') {
      presetModuleImport = presetItem;
    } else {
      [presetModuleImport, presetOptions] = presetItem;
    }
    const presetName = resolveModuleName(
      presetModuleImport,
      presetRequire,
      'preset',
    );

    const presetPath = presetRequire.resolve(presetName);
    const presetModule = (await loadFreshModule(
      presetPath,
    )) as ImportedPresetModule;

    const presetFunction = presetModule.default ?? presetModule;

    return presetFunction(context, presetOptions);
  }

  const loadedPresets = await Promise.all(presets.map(loadPreset));

  const plugins = loadedPresets.flatMap((preset) => preset.plugins ?? []);
  const themes = loadedPresets.flatMap((preset) => preset.themes ?? []);

  return {plugins, themes};
}
