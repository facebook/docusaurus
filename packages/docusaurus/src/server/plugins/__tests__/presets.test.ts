/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {loadPresets} from '../presets';
import type {LoadContext} from '@docusaurus/types';

describe('loadPresets', () => {
  it('no presets', async () => {
    const context = {
      siteConfigPath: __dirname,
      siteConfig: {
        presets: [],
      },
    } as unknown as LoadContext;
    const presets = await loadPresets(context);
    expect(presets).toMatchInlineSnapshot(`
      {
        "plugins": [],
        "themes": [],
      }
    `);
  });

  it('cjs form', async () => {
    const context = {
      siteConfigPath: __dirname,
      siteConfig: {
        presets: [
          path.join(__dirname, '__fixtures__/presets/preset-plugins.cjs.js'),
        ],
      },
    } as LoadContext;
    const presets = await loadPresets(context);
    expect(presets).toMatchSnapshot();
  });

  it('esm form', async () => {
    const context = {
      siteConfigPath: __dirname,
      siteConfig: {
        presets: [
          path.join(__dirname, '__fixtures__/presets/preset-plugins.esm.js'),
        ],
      },
    } as LoadContext;
    const presets = await loadPresets(context);
    expect(presets).toMatchSnapshot();
  });

  it('ts form', async () => {
    const context = {
      siteConfigPath: __dirname,
      siteConfig: {
        presets: [
          path.join(__dirname, '__fixtures__/presets/preset-plugins.ts'),
        ],
      },
    } as LoadContext;
    const presets = await loadPresets(context);
    expect(presets).toMatchSnapshot();
  });

  it('string form composite', async () => {
    const context = {
      siteConfigPath: __dirname,
      siteConfig: {
        presets: [
          path.join(__dirname, '__fixtures__/presets/preset-plugins.cjs.js'),
          path.join(__dirname, '__fixtures__/presets/preset-themes.js'),
        ],
      },
    } as LoadContext;
    const presets = await loadPresets(context);
    expect(presets).toMatchSnapshot();
  });

  it('array form', async () => {
    const context = {
      siteConfigPath: __dirname,
      siteConfig: {
        presets: [
          [path.join(__dirname, '__fixtures__/presets/preset-plugins.cjs.js')],
        ],
      },
    } as unknown as LoadContext;
    const presets = await loadPresets(context);
    expect(presets).toMatchSnapshot();
  });

  it('array form with options', async () => {
    const context = {
      siteConfigPath: __dirname,
      siteConfig: {
        presets: [
          [
            path.join(__dirname, '__fixtures__/presets/preset-plugins.cjs.js'),
            {docs: {path: '../'}},
          ],
        ],
      },
    } as unknown as LoadContext;
    const presets = await loadPresets(context);
    expect(presets).toMatchSnapshot();
  });

  it('array form composite', async () => {
    const context = {
      siteConfigPath: __dirname,
      siteConfig: {
        presets: [
          [
            path.join(__dirname, '__fixtures__/presets/preset-plugins.cjs.js'),
            {docs: {path: '../'}},
          ],
          [
            path.join(__dirname, '__fixtures__/presets/preset-themes.js'),
            {algolia: {trackingID: 'foo'}},
          ],
        ],
      },
    } as unknown as LoadContext;
    const presets = await loadPresets(context);
    expect(presets).toMatchSnapshot();
  });

  it('mixed form', async () => {
    const context = {
      siteConfigPath: __dirname,
      siteConfig: {
        presets: [
          [
            path.join(__dirname, '__fixtures__/presets/preset-plugins.cjs.js'),
            {docs: {path: '../'}},
          ],
          path.join(__dirname, '__fixtures__/presets/preset-themes.js'),
        ],
      },
    } as LoadContext;
    const presets = await loadPresets(context);
    expect(presets).toMatchSnapshot();
  });

  it('mixed form with themes', async () => {
    const context = {
      siteConfigPath: __dirname,
      siteConfig: {
        presets: [
          [
            path.join(__dirname, '__fixtures__/presets/preset-plugins.cjs.js'),
            {docs: {path: '../'}},
          ],
          false,
          null,
          undefined,
          path.join(__dirname, '__fixtures__/presets/preset-themes.js'),
          path.join(__dirname, '__fixtures__/presets/preset-mixed.js'),
        ],
      },
    } as LoadContext;
    const presets = await loadPresets(context);
    expect(presets).toMatchSnapshot();
  });
});
