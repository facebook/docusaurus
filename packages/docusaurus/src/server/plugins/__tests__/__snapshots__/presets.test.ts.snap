// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadPresets array form 1`] = `
{
  "plugins": [
    [
      "@docusaurus/plugin-content-docs",
      undefined,
    ],
    [
      "@docusaurus/plugin-content-blog",
      undefined,
    ],
  ],
  "themes": [],
}
`;

exports[`loadPresets array form composite 1`] = `
{
  "plugins": [
    [
      "@docusaurus/plugin-content-docs",
      {
        "path": "../",
      },
    ],
    [
      "@docusaurus/plugin-content-blog",
      undefined,
    ],
  ],
  "themes": [
    [
      "@docusaurus/theme-live-codeblock",
      undefined,
    ],
    [
      "@docusaurus/theme-algolia",
      {
        "trackingID": "foo",
      },
    ],
  ],
}
`;

exports[`loadPresets array form with options 1`] = `
{
  "plugins": [
    [
      "@docusaurus/plugin-content-docs",
      {
        "path": "../",
      },
    ],
    [
      "@docusaurus/plugin-content-blog",
      undefined,
    ],
  ],
  "themes": [],
}
`;

exports[`loadPresets cjs form 1`] = `
{
  "plugins": [
    [
      "@docusaurus/plugin-content-docs",
      undefined,
    ],
    [
      "@docusaurus/plugin-content-blog",
      undefined,
    ],
  ],
  "themes": [],
}
`;

exports[`loadPresets esm form 1`] = `
{
  "plugins": [
    [
      "@docusaurus/plugin-content-docs",
      undefined,
    ],
    [
      "@docusaurus/plugin-content-blog",
      undefined,
    ],
  ],
  "themes": [],
}
`;

exports[`loadPresets mixed form 1`] = `
{
  "plugins": [
    [
      "@docusaurus/plugin-content-docs",
      {
        "path": "../",
      },
    ],
    [
      "@docusaurus/plugin-content-blog",
      undefined,
    ],
  ],
  "themes": [
    [
      "@docusaurus/theme-live-codeblock",
      undefined,
    ],
    [
      "@docusaurus/theme-algolia",
      undefined,
    ],
  ],
}
`;

exports[`loadPresets mixed form with themes 1`] = `
{
  "plugins": [
    [
      "@docusaurus/plugin-content-docs",
      {
        "path": "../",
      },
    ],
    [
      "@docusaurus/plugin-content-blog",
      undefined,
    ],
    [
      "@docusaurus/plugin-test",
      undefined,
    ],
    false,
  ],
  "themes": [
    [
      "@docusaurus/theme-live-codeblock",
      undefined,
    ],
    [
      "@docusaurus/theme-algolia",
      undefined,
    ],
    [
      "@docusaurus/theme-classic",
      undefined,
    ],
    null,
  ],
}
`;

exports[`loadPresets string form composite 1`] = `
{
  "plugins": [
    [
      "@docusaurus/plugin-content-docs",
      undefined,
    ],
    [
      "@docusaurus/plugin-content-blog",
      undefined,
    ],
  ],
  "themes": [
    [
      "@docusaurus/theme-live-codeblock",
      undefined,
    ],
    [
      "@docusaurus/theme-algolia",
      undefined,
    ],
  ],
}
`;

exports[`loadPresets ts form 1`] = `
{
  "plugins": [
    [
      "@docusaurus/plugin-test",
      undefined,
    ],
    false,
  ],
  "themes": [
    [
      "@docusaurus/theme-classic",
      undefined,
    ],
    null,
  ],
}
`;
